File: BaekJoon Online Judge/C++/Step15-Stack/1874.cpp
//
//  1874.cpp
//  Step15-Stack
//
//  Created by 김예빈 on 2019. 9. 20..
//  Copyright © 2019년 김예빈. All rights reserved.
//

#include <iostream>
#include <stack>
#include <vector>
using namespace std;

int main(int argc, const char * argv[]) {
    cin.tie(NULL);
    ios::sync_with_stdio(false);
    
    int n, cur = 0, cnt = 1;
    bool check = true;
    stack<int> st;
    vector<char> result;
    vector<char>::iterator iter;
    
    cin >> n;
    
    int num[n];
    
    for(int i = 0; i < n; i++)
        cin >> num[i];
    
    while(1) {
        for(int i = cnt; i <= num[cur]; i++) {
            st.push(i);
            result.push_back('+');
            cnt++;
        }
        
        if(st.top() == num[cur] && !st.empty()) {
            st.pop();
            result.push_back('-');
        } else {
            check = false;
            break;
        }
        
        cur++;
        if(cur == n) break;
    }
    
    if(check) {
        for(iter = result.begin(); iter != result.end(); iter++)
            cout << *iter << "\n";
    } else {
        cout << "NO";
    }
    
    return 0;
}

File: BaekJoon Online Judge/Swift/Step17-Stacks/1874.swift
//
//  1874.swift
//  Step17-Stacks
//
//  Created by Yebin Kim on 2019/12/31.
//  Copyright © 2019 김예빈. All rights reserved.
//

import Foundation

public struct Stack<T> {
    private var stack = Array<T>()

    public mutating func push(_ element: T) {
        stack.append(element)
    }

    public mutating func pop() {
        stack.removeLast()
    }

    public var top: T? {
        return stack.last
    }

    public var isEmpty: Bool {
        return stack.isEmpty
    }
}

extension Stack: Sequence {
    public func makeIterator() -> AnyIterator<T> {
        return AnyIterator(IndexingIterator(_elements: stack.lazy))
    }
}

let N = Int(readLine()!)!
var num = [Int]()
var st = Stack<Int>()
var answer = Stack<Character>()
var cnt = 1, cur = 0

for _ in 0..<N {
    num.append(Int(readLine()!)!)
}

while true {
    while cnt <= num[cur] {
        st.push(cnt);
        answer.push("+");
        cnt += 1;
    }
    
    if st.top == num[cur] && !st.isEmpty {
        st.pop();
        answer.push("-");
    } else {
        print("NO")
        break;
    }
    
    cur += 1;
    
    if cur == N {
        for e in answer {
            print(e)
        }
        
        break
    }
}
